File: source/prfs_api_server/src/apis/prfs_circuit_drivers.rs
use prfs_api_error_codes::PRFS_API_ERROR_CODES;
use prfs_axum_lib::axum::{extract::State, http::StatusCode, Json};
use prfs_axum_lib::resp::ApiResponse;
use prfs_common_server_state::ServerState;
use prfs_db_interface::prfs;
use prfs_entities::prfs_api::{
    GetPrfsCircuitDriverByDriverIdRequest, GetPrfsCircuitDriverByDriverIdResponse,
    GetPrfsCircuitDriversRequest, GetPrfsCircuitDriversResponse,
};
use std::sync::Arc;

pub async fn get_prfs_circuit_drivers(
    State(state): State<Arc<ServerState>>,
    Json(input): Json<GetPrfsCircuitDriversRequest>,
) -> (StatusCode, Json<ApiResponse<GetPrfsCircuitDriversResponse>>) {
    let pool = &state.clone().db2.pool;
    let prfs_circuit_drivers = match prfs::get_prfs_circuit_drivers(&pool).await {
        Ok(d) => d,
        Err(err) => {
            let resp = ApiResponse::new_error(&PRFS_API_ERROR_CODES.UNKNOWN_ERROR, err.to_string());
            return (StatusCode::BAD_REQUEST, Json(resp));
        }
    };

    let resp = ApiResponse::new_success(GetPrfsCircuitDriversResponse {
        page_idx: input.page_size,
        prfs_circuit_drivers,
    });
    return (StatusCode::OK, Json(resp));
}

pub async fn get_prfs_circuit_driver_by_driver_id(
    State(state): State<Arc<ServerState>>,
    Json(input): Json<GetPrfsCircuitDriverByDriverIdRequest>,
) -> (
    StatusCode,
    Json<ApiResponse<GetPrfsCircuitDriverByDriverIdResponse>>,
) {
    let pool = &state.clone().db2.pool;
    let prfs_circuit_driver =
        match prfs::get_prfs_circuit_driver_by_circuit_driver_id(&pool, &input.circuit_driver_id)
            .await
        {
            Ok(d) => d,
            Err(err) => {
                let resp =
                    ApiResponse::new_error(&PRFS_API_ERROR_CODES.UNKNOWN_ERROR, err.to_string());
                return (StatusCode::BAD_REQUEST, Json(resp));
            }
        };

    let resp = ApiResponse::new_success(GetPrfsCircuitDriverByDriverIdResponse {
        prfs_circuit_driver,
    });
    return (StatusCode::OK, Json(resp));
}

File: source/prfs_api_js/src/prfs_circuit_drivers.ts
import { PrfsApiResponse } from "./types";
import { PrfsCircuitDriver } from "@taigalabs/prfs-entities/bindings/PrfsCircuitDriver";
import { GetPrfsCircuitDriversRequest } from "@taigalabs/prfs-entities/bindings/GetPrfsCircuitDriversRequest";
import { GetPrfsCircuitDriversResponse } from "@taigalabs/prfs-entities/bindings/GetPrfsCircuitDriversResponse";
import { GetPrfsCircuitDriverByDriverIdRequest } from "@taigalabs/prfs-entities/bindings/GetPrfsCircuitDriverByDriverIdRequest";
import { GetPrfsCircuitDriverByDriverIdResponse } from "@taigalabs/prfs-entities/bindings/GetPrfsCircuitDriverByDriverIdResponse";

import { api } from "./utils";

export async function getPrfsCircuitDrivers(req: GetPrfsCircuitDriversRequest) {
  return (await api({
    path: `get_prfs_circuit_drivers`,
    req,
  })) as PrfsApiResponse<GetPrfsCircuitDriversResponse>;
}

export async function getPrfsCircuitDriverByDriverId(req: GetPrfsCircuitDriverByDriverIdRequest) {
  return (await api({
    path: `get_prfs_circuit_driver_by_driver_id`,
    req,
  })) as PrfsApiResponse<GetPrfsCircuitDriverByDriverIdResponse>;
}
